File: vuln-sandbox/go/cwe-22/path-traversal-safe.go
// SAFE: go-os-open-concat — path.Clean and containment check prevent directory traversal
// Rule: go-os-open-concat | CWE-22 | Expected: TrueNegative

package main

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
)

const uploadDir = "/uploads"

func fileHandler(w http.ResponseWriter, r *http.Request) {
	filename := r.URL.Query().Get("file")

	// SAFE: clean the path and verify it stays within the allowed directory
	cleanPath := filepath.Clean(filepath.Join(uploadDir, filename))
	if !strings.HasPrefix(cleanPath, uploadDir+string(os.PathSeparator)) {
		http.Error(w, "invalid file path", http.StatusBadRequest)
		return
	}

	f, err := os.Open(cleanPath)
	if err != nil {
		http.Error(w, "file not found", http.StatusNotFound)
		return
	}
	defer f.Close()

	data, err := io.ReadAll(f)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	fmt.Fprintf(w, "%s", data)
}

func main() {
	http.HandleFunc("/file", fileHandler)
	http.ListenAndServe(":8080", nil)
}

File: vuln-sandbox/typescript/cwe-22/path-traversal-safe.ts
// SAFE: path-traversal — path normalization and containment check prevent traversal (TypeScript)
// Rule: InputPathTraversal | CWE-22 | Expected: TrueNegative

import express, { Request, Response } from 'express';
import path from 'path';
import fs from 'fs';

const app = express();
const UPLOAD_DIR: string = path.resolve('/uploads');

app.get('/file', (req: Request, res: Response): void => {
  const filename = req.query.file as string;

  // SAFE: resolve the full path and verify it stays within the allowed directory
  const filePath = path.resolve(UPLOAD_DIR, path.basename(filename));
  if (!filePath.startsWith(UPLOAD_DIR + path.sep)) {
    res.status(400).json({ error: 'Invalid file path' });
    return;
  }

  if (!fs.existsSync(filePath)) {
    res.status(404).json({ error: 'File not found' });
    return;
  }

  res.sendFile(filePath);
});

app.listen(3000);
